daemon/src/collab_settlement_taker.rs

```diff
@@ -0,0 +1,194 @@
+                msg: wire::taker_to_maker::Settlement::Initiate { sig_taker: sig },
+            })
+            .await?;
+
+        Ok(CollaborativeSettlement::new(
+            tx,
+            dlc.script_pubkey_for(self.cfd.role()), // TODO: Hardcode role to Taker?
+            self.proposal.price,
+        )?)
+    }
+
+    async fn handle_rejected(&mut self) -> Result<()> {
+        let order_id = self.cfd.order.id;
+
+        tracing::info!(%order_id, "Settlement proposal got rejected");
+
+        self.update_proposal(None).await?;
+
+        Ok(())
+    }
+
+    async fn update_proposal(
+        &mut self,
+        proposal: Option<(SettlementProposal, SettlementKind)>,
+    ) -> Result<()> {
+        self.projection
+            .send(projection::UpdateSettlementProposal {
+                order: self.cfd.order.id,
+                proposal,
+            })
+            .await?;
+
+        Ok(())
+    }
+
+    async fn complete(&mut self, completed: Completed, ctx: &mut xtra::Context<Self>) {
+        let _ = self.on_completed.send(completed).await;
+
+        ctx.stop();
+    }
+}
+
+#[async_trait]
+impl xtra::Actor for Actor {
+    async fn started(&mut self, ctx: &mut xtra::Context<Self>) {
+        let this = ctx.address().expect("get address to ourselves");
+
+        if let Err(e) = self.propose(this).await {
+            self.complete(
+                Completed::Failed {
+                    order_id: self.cfd.order.id,
+                    error: e,
+                },
+                ctx,
+            )
+            .await;
+        }
+    }
+
+    async fn stopping(&mut self, ctx: &mut xtra::Context<Self>) -> xtra::KeepRunning {
+        // inform the connection actor that we stopping so it can GC the address from the hashmap
+        let me = ctx.address().expect("we are still alive");
+        let _ = self.connection.send(Stopping { me }).await;
+
+        xtra::KeepRunning::StopAll
+    }
+}
+
+pub enum Completed {
+    Confirmed {
+        order_id: OrderId,
+        settlement: CollaborativeSettlement,
+    },
+    Rejected {
+        order_id: OrderId,
+    },
+    Failed {
+        order_id: OrderId,
+        error: anyhow::Error,
+    },
+}
+
+#[xtra_productivity]
+impl Actor {
+    async fn handle(
+        &mut self,
+        msg: wire::maker_to_taker::Settlement,
+        ctx: &mut xtra::Context<Self>,
+    ) {
+        let order_id = self.cfd.order.id;
+
+        let completed = match msg {
+            wire::maker_to_taker::Settlement::Confirm => match self.handle_confirmed().await {
+                Ok(settlement) => Completed::Confirmed {
+                    settlement,
+                    order_id,
+                },
+                Err(e) => Completed::Failed { error: e, order_id },
+            },
+            wire::maker_to_taker::Settlement::Reject => {
+                if let Err(e) = self.handle_rejected().await {
+                    Completed::Failed { error: e, order_id }
```

Comment:
:thinking: A little annoying that handling a rejection can fail.

Reply:
We could actually make it non-fallible. If the `projection::Actor` goes away (which is the only error case of `send`), we don't need to fail the entire settlement.